### PR TITLE
Better HTML report model

### PR DESCRIPTION
**Pull request by lptr**
Vue.js allows us to better model things. Let's make use of it.

[test-report.html.zip](https://github.com/gradle/gradle-profiler/files/5218974/test-report.html.zip)

### CODE REVIEW

File: src/main/resources/org/gradle/profiler/report/report-template.html [hunk 1]
```diff
@@ -148,12 +142,21 @@
...
+    <span>Gradle Profiler version {{ benchmarkResult.environment.profilerVersion }}</span>
 </footer>
 </div>
 <script>
 Array.prototype.unique = function() { return [...new Set(this)]; };
 
+// Initialize sample config
+benchmarkResult.scenarios.forEach((scenario, scenarioIndex, scenarios) => {
+    scenario.samples.forEach((sample, sampleIndex, samples) => {
+        sample.color = `hsl(${scenarioIndex * 360 / scenarios.length}, ${100 - 80 * sampleIndex / samples.length}%, ${30 + 40 * sampleIndex / samples.length}%)`;
```
Comment (wolfs) [inline src/main/resources/org/gradle/profiler/report/report-template.html:155]: I find the shades of red/green very hard to distinguish: 
image

Do you have any better ideas what colors to use?

File: src/main/resources/org/gradle/profiler/report/report-template.html [hunk 2]
```diff
@@ -190,132 +193,118 @@
...
-    let samples = $("#sample").val();
-    chart.data.datasets = convertData(samples, sorted);
-    chart.update();
-}
-
-let ctx = document.getElementById('samples').getContext('2d');
-let maxMeasuredIterations = Math.max(...(benchmarkResult.scenarios.map(scenario => iterations(scenario, Phase.MEASURE).length)));
-let chart = new Chart(ctx, {
-    type: 'line',
+var app = new Vue({
```
Comment (wolfs) [inline src/main/resources/org/gradle/profiler/report/report-template.html:218]:
```suggestion
const app = new Vue({
```

No need to leave this mutable.

File: src/main/resources/org/gradle/profiler/report/report-template.html [hunk 3]
```diff
@@ -148,12 +142,21 @@
 </div>
 </div>
 <footer>
-    <span>Gradle Profiler version {{ environment.profilerVersion }}</span>
+    <span>Gradle Profiler version {{ benchmarkResult.environment.profilerVersion }}</span>
 </footer>
 </div>
 <script>
 Array.prototype.unique = function() { return [...new Set(this)]; };
```
Comment (wolfs) [inline src/main/resources/org/gradle/profiler/report/report-template.html:150]: The `unique` function is now unused.

File: src/main/resources/org/gradle/profiler/report/report-template.html [hunk 4]
```diff
@@ -190,132 +194,118 @@
...
-        if (this.colors.length) {
-            return this.colors.shift();
-        } else {
-            return "#8B899A";
-        }
-    }
-};
-
 function convertScenarioToChart(scenario, sample, sorted, color) {
     var data = iterations(scenario, Phase.MEASURE)
```
Comment (wolfs) [inline src/main/resources/org/gradle/profiler/report/report-template.html:198]:
```suggestion
    const data = iterations(scenario, Phase.MEASURE)
```

**Review by wolfs — CHANGES_REQUESTED** [review]
Nice! I added some comments.

File: src/main/resources/org/gradle/profiler/report/report-template.html (hunk 2 above)
Comment (lptr) [reply src/main/resources/org/gradle/profiler/report/report-template.html]: I've fixed this and others.

File: src/main/resources/org/gradle/profiler/report/report-template.html (hunk 3 above)
Comment (lptr) [reply src/main/resources/org/gradle/profiler/report/report-template.html]: I removed it and moved the utility functions into `Array` prototype functions.

File: src/main/resources/org/gradle/profiler/report/report-template.html (hunk 4 above)
Comment (lptr) [reply src/main/resources/org/gradle/profiler/report/report-template.html]: Done.

File: src/main/resources/org/gradle/profiler/report/report-template.html (hunk 1 above)
Comment (lptr) [reply src/main/resources/org/gradle/profiler/report/report-template.html]: Not sure how important it is to distinguish the different shades of green (=different build operations for the same experiment), it seems more important to distinguish the green from the red (same build operations for different experiments).

Comment (lptr) [reply src/main/resources/org/gradle/profiler/report/report-template.html]: That said I'm open to suggestions.

**Review by wolfs — APPROVED** [review]
LGTM! I still wonder if we can do something about the colors. Let's merge this PR and discuss to see if we can come up with something.

**Pull request merged**